aesara/link/numba/dispatch/basic.py

```diff
@@ -515,6 +514,91 @@ def numba_funcify_IncSubtensor(op, node, **kwargs):
     return numba_njit(incsubtensor_fn)
 
 
+@numba_funcify.register(AdvancedIncSubtensor1)
+def numba_funcify_AdvancdIncSubtensor1(op, node, **kwargs):
+    inplace = op.inplace
+    set_instead_of_inc = op.set_instead_of_inc
+
+    # We can specialize if we know the indexes beforhand
+    # and precompute bounds checks.
+    idxs = node.inputs[2]
+    assert np.dtype(idxs.type.dtype).kind in ["i", "u"]
+    if isinstance(idxs, TensorConstant) and len(idxs.data) > 0:
+        idxs = idxs.data
+        idxs_dtype = np.dtype(idxs.dtype)
+        min_idx = idxs.min()
+        max_idx = idxs.max()
+        # Convert to unsigned index if possible to speed up indexing
+        if min_idx >= 0 and idxs_dtype.kind == "i":
+            idxs = idxs.astype(np.dtype(f"u{idxs_dtype.itemsize}"))
+
+        msg_pos = f"Bad index: {max_idx}"
+        msg_neg = f"Bad index: {min_idx}"
+
+        @numba_njit
+        def check(z, vals):
+            if max_idx >= len(z):
+                raise IndexError(msg_pos)
+            if max(-min_idx, 0) > len(z):
+                raise IndexError(msg_neg)
+            if len(idxs) != len(vals):
+                raise ValueError("Incompatible dimensions during indexing")
```

Comment:
We generally try to avoid manual checks at the Numba level.  Partly because the same checks can occur downstream at the Numba level.  For instance, if the values are constants, then some such checks occur at compile-time (e.g. during constant folding-like optimization passes), and at run-time in the code generated by `zip`, `for`, and `z[idx]`.  

Also, we want to honor the Numba-level options for disabling things like bounds checks.  If we add similar checks of our own, then we can end up effectively overriding such options.

There are definitely cases in which manual checks are warranted, but it's not clear to me whether or not this is one of those cases.

Reply:
The point of that is that if we do it manually, we can always set `boundscheck=False` for that function without any safety issues (if the indices are known at compile time, as indicated in the code.)
I checked that this cleans up the generated assembly quite a bit, and also speeds up the execution of the function.
I've seen a lot of models where a big share of the execution time is in this op, so even smaller performance improvements matter.

I think even in general manual checks might be a good idea in many cases, often the optimizer will then optimize away the later checks and we end up with cleaner and faster asm code. Generally, the fewer checks happen within the loop the better.